draft-ietf-tls-external-psk-importer.md

```diff
@@ -138,14 +178,46 @@ EPSKs may be imported for early data use if they are bound to protocol settings
 otherwise be required for early data with normal (ticket-based PSK) resumption. Minimally, that means ALPN,
 QUIC transport settings, etc., must be provisioned alongside these EPSKs.
 
+##Client ID {#client-id-description}
+This is to prevent Selfie-style reflections.
+This is only necessary in scenarios where more than two actors use the same key, including the case where a single agent will complete PSK handshakes as both the client and the server using the same key.
+The Selfie attack {{!Selfie}} abuses an underlying assumption of TLS, that only one client and one server know a PSK.
+A pair of agents that both act as both client *and* server, and use the same PSK in both roles violate this assumption.
+In this case, there are two clients and two servers that know the PSK, since each endpoint acts as both client and server.
+The agents are thus both vulnerable to reflection attacks where the attacker forces the agent to act as both client and server in a single connection.
+This is because neither agent can distinguish between itself and its peer.
+Recall both servers have the exact same configuration.
+Therefore even though the session specific parameters change for each session, the client cannot distinguish between a response reflected back on it by the attacker and an honest response from its peer.
+For each parameter, both servers either return the value for every session, or they return a different value for every session.
+By hashing a distinguishing `client_id` value into the PSK binder, a server that received a reflected `ClientHello` would be unable to verify the binder, and would reject the connection.
+
+For example, distinguishing strings for agents in peer-to-peer IoT device deployments could be a MAC address.
+Similarly, distinguishing strings for agents in mesh VM deployments could be namespaces.
+The decision to use this field and what it should contain MUST be agreed OOB.
+Note that, critically, this value is never sent on the wire, as this would identify the client even to passive adverseries.
+
+##Prior Contexts {#prior-contexts-description}
+In the standard case this list will be empty because the TLS connection will not be wrapped in a prior security context.
+However, if the OOB PSK was established through a protocol, or series of protocols, that provide a continuing security context, then including the channel binding for these contexts, as well as any keys established, binds the security contexts together.
+This makes it easier to reason formally about the exact properties are provided by the combined series of protocols.
```

Comment:
I guess you are concerned about a situation where two different key inputs result in the same session keys (and thus the same channel binding)? Then why not define the channel binding as including the hash of the previous transcript?

Requiring previous keys to be included here might actually weaken security as they would have to be kept around for a longer time. (That could be mitigated by precomputing the context, but that might not always be possible.)

Reply:
Correct. 
The reason I don't define the channel binding in that way is because it's neither a sufficient nor a necessary condition, even though including it is a very common design pattern. 
A channel binding based solely on the (public) transcript is not necessarily secure, because some security critical parameters, e.g. the keys being used, are secret. 
Consider the attack in [Cremers et al.](https://tls13tamarin.github.io/TLS13Tamarin/docs/tls13tamarin.pdf). 
The transcripts of the second session are both identical, but they agree different session keys. 

Transcripts are also not necessary. [Bhargavan et al.](https://www.ndss-symposium.org/wp-content/uploads/2017/09/08_5.pdf) showed that you can build secure channel bindings without using a transcript. They provide a list of what is needed in Section IV. 

I agree that keeping session keys around long term is bad practice, but that isn't the intention here. 
This is for if you want to run a compound protocol, i.e. nest one protocol inside the security context of another.
The idea is that you run all protocols in quick succession, as if they were one extended protocol. 
Because we are building a nested structure, and you keep session keys until the end of the session, if you want the security context of protocol A nested inside the security context of protocol B, you run B, keeping the keys, then run A, then delete any keys for A, then delete any keys for B. `[ ( ) ]` 
The design intentionally requires you to prove that you have access to all the keys during the establishment of the innermost session. 
Access to the channel binding is insufficient, because channel bindings are designed such that you can make them public without damaging security. 
If you have to keep your channel binding secet it's not a channel binding, it's a key. 

TL;DR: requiring the keys is a necessary condition, the channel binding is insufficient.